Build fails with newer versions of grpc

I'm running into an installation error on a Mac M1. I think the problem comes with using the latest version of `grpc` (1.67.1) and it's related to #71.

```
install.packages("bigrquerystorage")
...
In file included from bqs.cpp:7:
In file included from ./google/cloud/bigquery/storage/v1/stream.pb.h:26:
In file included from /opt/homebrew/Cellar/protobuf/28.3/include/google/protobuf/generated_message_reflection.h:25:
In file included from /opt/homebrew/Cellar/protobuf/28.3/include/google/protobuf/descriptor.h:45:
In file included from /opt/homebrew/Cellar/abseil/20240722.0/include/absl/container/btree_map.h:59:
In file included from /opt/homebrew/Cellar/abseil/20240722.0/include/absl/container/internal/btree.h:67:
In file included from /opt/homebrew/Cellar/abseil/20240722.0/include/absl/container/internal/layout.h:199:
In file included from /opt/homebrew/Cellar/abseil/20240722.0/include/absl/strings/str_cat.h:110:
/opt/homebrew/Cellar/abseil/20240722.0/include/absl/strings/numbers.h:186:5: warning: 'diagnose_if' is a clang extension [-Wgcc-compat]
  186 |     ABSL_INTERNAL_NEED_MIN_SIZE(buffer, kFastToBufferSize) {
      |     ^
/opt/homebrew/Cellar/abseil/20240722.0/include/absl/base/macros.h:181:18: note: expanded from macro 'ABSL_INTERNAL_NEED_MIN_SIZE'
  181 |   __attribute__((diagnose_if(__builtin_object_size(Obj, 0) < N, \
      |                  ^
bqs.cpp:20:22: error: unknown type name 'gpr_log_func_args'
   20 | void bqs_default_log(gpr_log_func_args* args) {
      |                      ^
bqs.cpp:32:3: error: use of undeclared identifier 'gpr_set_log_verbosity'; did you mean 'bqs_set_log_verbosity'?
   32 |   gpr_set_log_verbosity(static_cast<gpr_log_severity>(severity));
      |   ^~~~~~~~~~~~~~~~~~~~~
      |   bqs_set_log_verbosity
bqs.cpp:26:6: note: 'bqs_set_log_verbosity' declared here
   26 | void bqs_set_log_verbosity(int severity) {
      |      ^
12 warnings and 2 errors generated.
make: *** [bqs.o] Error 1
ERROR: compilation failed for package ‘bigrquerystorage’
```

These old log functions seem to now be deprecated by `grpc`. I noticed [here](https://github.com/grpc/grpc/commit/cf88319446645435035a5d3d06d19348a831661d) that you can see the removal of one of them and overall it seems `grpc` has made a move to absl logging.

--- a/src/bqs.cpp
+++ b/src/bqs.cpp
@@ -2,7 +2,27 @@
 #include <string>
 #include <vector>
 #include <grpc/grpc.h>
+
+#if __has_include(<grpcpp/version_info.h>)
+#include <grpcpp/version_info.h>
+#if GRPC_CPP_VERSION_MAJOR >= 1 & GRPC_CPP_VERSION_MINOR >= 67
+#define ABSL_LOGGING 1
+#endif
+#endif
+
+#ifndef ABSL_LOGGING
 #include <grpc/support/log.h>
+#endif
+
+#ifdef ABSL_LOGGING
+#include <absl/base/no_destructor.h>
+#include <absl/base/log_severity.h>
+#include <absl/log/globals.h>
+#include <absl/log/initialize.h>
+#include <absl/log/log_sink_registry.h>
+#include <absl/log/log_sink.h>
+#endif
+
 #include <grpcpp/grpcpp.h>
 #include "google/cloud/bigquery/storage/v1/stream.pb.h"
 #include "google/cloud/bigquery/storage/v1/storage.pb.h"
@@ -15,27 +35,60 @@ using google::cloud::bigquery::storage::v1::ReadSession;
 using google::cloud::bigquery::storage::v1::BigQueryRead;
 
 // -- Utilities and logging ----------------------------------------------------
-
 // Define a default logger for gRPC
+#ifndef ABSL_LOGGING
 void bqs_default_log(gpr_log_func_args* args) {
   Rcpp::Rcerr << args->message << std::endl;
 }
+#endif
+
+#ifdef ABSL_LOGGING
+class RLogSink : public absl::LogSink {
+public:
+  void Send(const absl::LogEntry& entry) override {
+    Rcpp::Rcerr << entry.text_message_with_prefix_and_newline() << std::endl;
+  }
+};
+#endif
 
 // Set gRPC verbosity level
 // [[Rcpp::export(rng=false)]]
 void bqs_set_log_verbosity(int severity) {
+
+#ifndef ABSL_LOGGING
   //-1 UNSET
   // 0 DEBUG
   // 1 INFO
   // 2 ERROR
   // 3 QUIET
   gpr_set_log_verbosity(static_cast<gpr_log_severity>(severity));
+#endif
+
+#ifdef ABSL_LOGGING
+  // -1 UNSET
+  //  0 INFO
+  //  1 WARNING
+  //  2 ERROR
+  //  3 FATAL
+  absl::SetMinLogLevel(static_cast<absl::LogSeverityAtLeast>(severity));
+#endif
+
 }
 
 // Set gRPC default logger
 // [[Rcpp::export(rng=false)]]
 void bqs_init_logger() {
+
+#ifndef ABSL_LOGGING
   gpr_set_log_function(bqs_default_log);
+#endif
+
+#ifdef ABSL_LOGGING
+  absl::InitializeLog();
+  static absl::NoDestructor<RLogSink> r_log_sink;
+  absl::AddLogSink(r_log_sink.get());
+#endif
+
   bqs_set_log_verbosity(2);
 }
 

